### ISSUE TITLE
Differences between PyTorch and Flax model predictions

### ISSUE THREAD

**Issue by JamesAllingham** [google/flax#963]
Hi all,

Here is a bit of a weird one that has me stumped!

### Problem you have encountered:

I've been trying to convert ResNet models trained in PyTorch to Flax and I've run into a strange issue. The PyTorch and Flax models tend to make the same class predictions but the class probabilities are slightly different. I've realised that each layer is producing slightly different outputs and that the errors slowly add up. In the case of a ResNet18, the difference in the final prediction is minimal but for a ResNet50 it can mean that the incorrect class is predicted! 

This discrepancy also occurs of a single dense layer (see my minimal reproducible example below).

### What you expected to happen:

I would expect that the same model architecture with the same weights and the same input would produce exactly the same output.

### Steps to reproduce:
Here is a short python script that demonstrates the problem:

```
from jax import numpy as jnp

from flax import linen
from flax.core import freeze

import torch

import numpy as np

torch.manual_seed(0)

torch_dense = torch.nn.Linear(2048, 10)
jax_dense = linen.Dense(10)

x = jnp.ones((2048,), dtype=np.float32)

jax_params = freeze({"params": {k if k == "bias" else "kernel": v.numpy().T for k, v in torch_dense.state_dict().items()}})
y_jax = jax_dense.apply(jax_params, x)

y_pytorch = torch_dense(torch.Tensor(np.ones((1, 2048), dtype=np.float32))).detach().numpy()[0]

print(f"2-norm of jax pred:\t {np.sqrt(np.mean((y_jax)**2)):.10f}")
print(f"2-norm of pytorch pred:\t {np.sqrt(np.mean((y_pytorch)**2)):.10f}")
print(f"2-norm of diff in preds: {np.sqrt(np.mean((y_jax-y_pytorch)**2)):.10f}")

print("Comparing the weight matrix...")
state = torch_dense.state_dict()["weight"]
numpy_state = state.numpy()
print(f"torch vs np:\t{torch.sqrt(torch.mean((state - numpy_state)**2)):.30f}")
print(f"np vs jax.np:\t{jnp.sqrt(jnp.mean((numpy_state - jnp.array(numpy_state))**2)):.30f}")
```
which outputs:
```
2-norm of jax pred:      0.5016138554
2-norm of pytorch pred:  0.5016137958
2-norm of diff in preds: 0.0000001995
Comparing the weight matrix...
torch vs np:    0.000000000000000000000000000000
np vs jax.np:   0.000000000000000000000000000000
```

My Python/Jax/Flax/etc. versions are:
* Python: 3.7.9 (default, Aug 31 2020, 12:42:55) [GCC 7.3.0] :: Anaconda, Inc. on linux
* jax: 0.2.9
* jaxlib: 0.1.59 (+cuda101)
* flax: 0.3.0
* torch: 1.6.0
* numpy: 1.17.2

EDIT: [Here](https://colab.research.google.com/drive/1oSp_9zooyrihaFi6AoqQ3k7lk6dIFunN?usp=sharing
) is a colab notebook for good measure :) 

Strangely the difference in predictions is slightly different:

```
2-norm of jax pred:	 0.5016137958
2-norm of pytorch pred:	 0.5016138554
2-norm of diff in preds: 0.0000004255
Comparing the weight matrix...
torch vs np:	0.000000000000000000000000000000
np vs jax.np:	0.000000000000000000000000000000
```

**Comment by levskaya** [comment]
@JamesAllingham - I don't think your colab is shared publicly.  Thanks for the detailed report!  Are you evaluating this on GPU 
 or CPU (or even cloud TPU?).

**Comment by levskaya** [comment]
You've actually seen class-label changes resulting from the propagated errors in a Resnet50 model?

**Comment by jheek** [comment]
For CPUs we cannot really guarantee much because XLA and PyTorch use completely different code paths.
As for GPUs this can also be complicated. XLA is by default stochastic on GPU so even running the same code twice can result in numerical differences.
Producing the same code would at least need XLA and PyTorch to end up calling the same CuDNN api's and use the same format for inputs and weights. However, XLA is allowed to optimize the permutation in which arrays are stored freely (and the permutation can affect CuDNNs choice of algorithm).

**Comment by JamesAllingham** [comment]
@levskaya I think I've fixed the sharing settings. Sorry about that. This was being evaluated on the CPU. I should clarify that the class label change I saw with ResNet-50 was for a random input, which is obviously outside of the distribution of the training data and may not be that relevant in practice. (I should also say that I can't rule out the possibility of a bug in my code which causing additional problems.) I'll update the colab notebook to include a ResNet-50 example. 

@jheek I see, that makes sense, thanks for the explanation. I would expect some small numerical differences between XLA and PyTorch. However, I wouldn't expect a difference in the class predictions. I'll set up a more realistic test in the colab notebook and see if I can reproduce the problem.

**Comment by JamesAllingham** [comment]
@jheek @levskaya here is an updated collab which shows the difference between predictions of a PyTorch pretrained (default implementation) ResNet50 (+ 18) and the (almost) default Flax implementations when making predictions with the PyTorch params: https://colab.research.google.com/drive/1oSp_9zooyrihaFi6AoqQ3k7lk6dIFunN?usp=sharing

As mentioned before, the ResNet50 can often get the class wrong. The ResNet18 also seems to be able to get the class wrong some of the time. I had originally thought that it happened less often due to less accumulation of errors but I am less sure now. 

I am still not 100% confident that these strange observations are not due to a bug in my code, so take everything I've said with a grain of salt :)

**Comment by n2cholas** [comment]
I've noticed that padding can cause differences between JAX and PyTorch, I recently ported ResNeSt weights from PyTorch to Flax and needed to manually control the padding of all my layers (instead of using `SAME` or `VALID`). I noticed the code you posted uses the default padding everywhere, so that may be one source of the differences. [Here](https://github.com/n2cholas/jax-resnet/commit/f52cfa07c996c404744e6708a5d53035dde1678f) is a diff containing the changes I had to make to ResNeSt to make it match the reference implementation.

**Comment by JamesAllingham** [comment]
Thanks, @n2cholas – the padding was indeed the issue! 

I've updated the notebook with specified padding and the predictions are now all the same. 

I'd be happy to make a PR to update the Flax example ResNet code with padding to match Pytorch if that is something that would be useful? I guess the question is whether or not PyTorch is doing this right or wrong?

**Comment by n2cholas** [comment]
Glad it helped @JamesAllingham!  

I don't think either Flax or PyTorch is "wrong", just different default conventions. I believe the [TensorFlow ResNet50](https://www.tensorflow.org/api_docs/python/tf/keras/applications/ResNet50) uses the same padding convention as Flax, while many recent ResNet advancements in PyTorch stick with the PyTorch default.
